// jwt-blacklist/dist/blacklist.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBlackList = exports.BlackList = void 0;
var jwt = __importStar(require("jsonwebtoken"));
var config_1 = require("./config");
var utils_1 = require("./utils");
var memory_filter_1 = require("./filter/memory-filter");
var redis_filter_1 = require("./filter/redis-filter");
var BlackList = /** @class */ (function () {
    function BlackList(config) {
        if (config === void 0) { config = config_1.defaultConfig; }
        this.config = __assign(__assign({}, config_1.defaultConfig), config);
        this.config.redisOptions = __assign(__assign({}, new config_1.RedisOptions()), this.config.redisOptions);
        if (this.config.storeType === config_1.STORE_TYPE.MEMORY) {
            this.filter = new memory_filter_1.MemoryFilter(this.config);
        }
        else {
            this.filter = new redis_filter_1.RedisFilter(this.config);
        }
    }
    BlackList.prototype.add = function (token, options) {
        return __awaiter(this, void 0, void 0, function () {
            var decoded, exp, now;
            return __generator(this, function (_a) {
                decoded = jwt.decode(token, options);
                if (!decoded) {
                    return [2 /*return*/, false];
                }
                if (typeof decoded === 'string' || decoded instanceof String) {
                    return [2 /*return*/, false];
                }
                exp = decoded.exp || 0;
                now = utils_1.nowInSeconds();
                if (exp < now) {
                    return [2 /*return*/, false];
                }
                return [2 /*return*/, this.filter.add(token, exp)];
            });
        });
    };
    BlackList.prototype.has = function (token, options) {
        return __awaiter(this, void 0, void 0, function () {
            var decoded, exp, now;
            return __generator(this, function (_a) {
                decoded = jwt.decode(token, options);
                if (!decoded) {
                    return [2 /*return*/, false];
                }
                if (typeof decoded === 'string' || decoded instanceof String) {
                    return [2 /*return*/, false];
                }
                exp = decoded.exp || 0;
                now = utils_1.nowInSeconds();
                if (exp < now) {
                    return [2 /*return*/, false];
                }
                return [2 /*return*/, this.filter.has(token, exp)];
            });
        });
    };
    BlackList.prototype.clear = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.filter.clear()];
            });
        });
    };
    BlackList.prototype.getFilter = function () {
        return this.filter;
    };
    return BlackList;
}());
exports.BlackList = BlackList;
function createBlackList(config) {
    if (config === void 0) { config = config_1.defaultConfig; }
    return __awaiter(this, void 0, void 0, function () {
        var blackList;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    blackList = new BlackList(config);
                    return [4 /*yield*/, blackList.getFilter().initialize()];
                case 1:
                    _a.sent();
                    return [2 /*return*/, blackList];
            }
        });
    });
}
exports.createBlackList = createBlackList;
