// jwt-blacklist/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultConfig = exports.Config = exports.RedisOptions = exports.STORE_TYPE = void 0;
var STORE_TYPE;
(function (STORE_TYPE) {
    STORE_TYPE["MEMORY"] = "memory";
    STORE_TYPE["REDIS"] = "redis";
})(STORE_TYPE = exports.STORE_TYPE || (exports.STORE_TYPE = {}));
var RedisOptions = /** @class */ (function () {
    function RedisOptions() {
        this.host = '127.0.0.1';
        this.port = 6379;
        this.key = 'jwt-blacklist';
    }
    return RedisOptions;
}());
exports.RedisOptions = RedisOptions;
var Config = /** @class */ (function () {
    function Config() {
        this.daySize = 10000;
        this.errorRate = 0.001;
        this.storeType = STORE_TYPE.MEMORY;
        this.redisOptions = new RedisOptions();
    }
    return Config;
}());
exports.Config = Config;
exports.defaultConfig = {
    daySize: 10000,
    errorRate: 0.001,
    storeType: STORE_TYPE.MEMORY,
    redisOptions: new RedisOptions(),
};
